tc_aws/storages/s3_storage.py

```diff
@@ -23,53 +27,146 @@ def __init__(self, context):
+        except BotoCoreError as err:
+            logger.exception('Unable to store crypto object: %s', err)
+            return None
+
+        logger.debug(
+            "Stored crypto at %s (security key: %s)",
+            crypto_path,
+            self.context.server.security_key,
+        )
+
+        return file_abspath
+
+    async def put_detector_data(self, path, data):
+        """
+        Stores detector data at given path
+        :param string path: Path to store the data at
+        :param string data:  Data to store
+        :return: Path where the data is stored
+        :rtype: string
+        """
+        file_abspath = self._normalize_path(path)
+
+        path = '%s.detectors.txt' % splitext(file_abspath)[0]
+
+        try:
+            await self._put_object(dumps(data).encode('utf-8'), path)
+        except BotoCoreError as err:
+            logger.exception('Unable to store detector data: %s', err)
+            return None
+
+        return file_abspath
+
+    async def get_crypto(self, path):
+        """
+        Retrieves crypto data at path
+        :param string path: Path to search for crypto data
         """
-        super(Storage, self).remove(self._normalize_path(path), callback)
+        file_abspath = self._normalize_path(path)
+        crypto_path = "%s.txt" % (splitext(file_abspath)[0])
 
-    @return_future
-    def get(self, path, callback):
+        try:
+            file_key = await self.storage.get(crypto_path)
+        except ClientError as err:
+            logger.warn("[STORAGE] s3 key not found at %s" % crypto_path)
+            return None
+
+        file_key = await file_key['Body'].read()
+
+        return file_key.decode('utf-8')
+
+    async def get_detector_data(self, path):
+        """
+        Retrieves detector data from storage
+        :param string path: Path where the data is stored
+        :param callable callback: Callback function for once the retrieval is done
+        """
+        file_abspath = self._normalize_path(path)
+        path = '%s.detectors.txt' % splitext(file_abspath)[0]
+
+        try:
+            file_key = await self.storage.get(path)
+        except ClientError:
+            return None
+
+        if not file_key or self.is_expired(file_key) or 'Body' not in file_key:
+            return None
+
+        return loads(await file_key['Body'].read())
+
+    async def get(self, path):
         """
         Gets data at path
         :param string path: Path for data
-        :param callable callback: Callback function for once the retrieval is done
         """
 
-        def parse_body(key):
-            if key is None or self._get_error(key) is not None:
-                callback(None)
-            else:
-                callback(key['Body'].read())
+        try:
+            file = await super(Storage, self).get(path)
+        except BotoCoreError:
+            return None
+
+        return await file['Body'].read()
```

Comment:
In the above method you check to see whether Body exists. Wouldn't it be better to have a `read_object` function that does this logic and just call it everywhere?

Reply:
Not sure. This forced me thinking on what to do if there is an exception while reading from s3. Should code catch all external exceptions, log them and return None? or should it throw read exceptions to callers?